### PR TITLE
AA-98: Prevent bundling UserOps that access another UserOp's sender

### PR DESCRIPTION
**Pull request by forshtat**

### CODE REVIEW

File: packages/bundler/src/modules/BundleManager.ts [hunk 1]
```diff
@@ -190,6 +197,18 @@ export class BundleManager {
         this.mempoolManager.removeUserOp(entry.userOp)
         continue
       }
+
+      for (const storageAddress of Object.keys(validationResult.storageMap)) {
+        if (
+          storageAddress.toLowerCase() !== entry.userOp.sender.toLowerCase() &&
+          knownSenders.includes(storageAddress.toLowerCase())
+        ) {
+          console.log('UserOperation read a storage of a different sender entity')
```
Comment (drortirosh) [inline packages/bundler/src/modules/BundleManager.ts:206]: don't leave comments, especially not one with no context.
e.g. can be debug( userop of ${entry.userop.sender} references  another sender ${storageAddress}

(also, missing a test)

Comment (drortirosh) [reply packages/bundler/src/modules/BundleManager.ts:206]: on the bundler repo, I'd expect a at least a minimal test. on the spec-test there should be an exhaustive list of tests.

Comment (forshtat) [reply packages/bundler/src/modules/BundleManager.ts]: > on the bundler repo, I'd expect a at least a minimal test. on the spec-test there should be an exhaustive list of tests.

There are barely any tests in a bundler repo. Also the tests appear to run against `hardhat` node, I don't have any way of testing this code without `go-ethereum` as far as I know. There is a corresponding test in the spec-test repo.

Comment (drortirosh) [reply packages/bundler/src/modules/BundleManager.ts]: please see the ValidationManager.test.ts
it contains a set of tests (not comprehensive, but still) for all basic rules.
(and yes, it requires geth)

Comment (forshtat) [reply packages/bundler/src/modules/BundleManager.ts]: Pushed a test for this specific rule.

File: packages/bundler/contracts/tests/cross_sender_state/TestFakeWalletToken.sol [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        balances[_owner] = balance;
+    }
+
+    function sudoSetAnotherWallet(TestFakeWalletToken _anotherWallet) public {
+        anotherWallet = _anotherWallet;
+    }
+
+    function validateUserOp(UserOperation calldata userOp, bytes32, uint256)
+    public override returns (uint256 validationData) {
+        if (userOp.callData.length == 20) {
```
Comment (drortirosh) [inline packages/bundler/contracts/tests/cross_sender_state/TestFakeWalletToken.sol:34]: I think the calldata is not needed.
the code can be:
```
require(anotherWallet.balanceOf(this) > 0 )
balances[anotherWallet] = 0
```
and "anotherWallet" should be the owner of the 2nd userop to submit.

Comment (forshtat) [reply packages/bundler/contracts/tests/cross_sender_state/TestFakeWalletToken.sol:34]: This does not really work as both UserOps are marked as accessing "known sender storage", so you would get an empty bundle. I `map` the list of all "known senders" before bundling as was suggested in the issue description.

File: packages/bundler/test/BundlerManager.test.ts [hunk 3]
```diff
@@ -66,4 +71,94 @@ describe('#BundlerManager', () => {
...
+        ...cEmptyUserOp,
+        sender: wallet2.address
+      }
+      await methodHandler.sendUserOperation(userOp1, entryPoint.address)
+      await methodHandler.sendUserOperation(userOp2, entryPoint.address)
+
+      const bundle = await bundleMgr.sendNextBundle()
+      await bundleMgr.handlePastEvents()
+      const mempool = bundleMgr.mempoolManager.getSortedForInclusion()
+
```
Comment (drortirosh) [inline packages/bundler/test/BundlerManager.test.ts:157]: 1. I reverted the fix, so this test will fail.
2. added here: 
```
      console.log('sender1=', userOp1.sender)
      console.log('sender2=', userOp2.sender)
      console.log((await entryPoint.queryFilter(entryPoint.filters.UserOperationEvent())).map(e=>[e.event, e.args.sender, e.args.nonce, e.args.success]))
```
and I can see that both userops passed and got included, so there is no failure to detect..

**Review by drortirosh — APPROVED** [review]

**Pull request merged**